### ISSUE TITLE
Incorrect gradient for region with arg used as free variable

### ISSUE THREAD

**Issue by joelberkeley** [EnzymeAD/Enzyme-JAX#3204]
I'm seeing what I believe is an incorrect gradient for a `grad` in a `while` body, where the `grad` captures a variable.

Input test.mlir
```
module @root {
  func.func @main() -> tensor<f64> {
    %cst = stablehlo.constant dense<2.000000e-01> : tensor<f64>
    %0 = stablehlo.while(%iterArg = %cst) : tensor<f64>
    cond {
      %cst_0 = stablehlo.constant dense<1.000000e+00> : tensor<f64>
      %1 = stablehlo.compare LT, %iterArg, %cst_0 : (tensor<f64>, tensor<f64>) -> tensor<i1>
      stablehlo.return %1 : tensor<i1>
    } do {
      %cst_0 = stablehlo.constant dense<1.000000e+00> : tensor<f64>
      %1 = enzyme.autodiff_region(%iterArg, %cst_0) {
      ^bb0(%arg0: tensor<f64>):
        %cst_1 = stablehlo.constant dense<3.000000e+00> : tensor<f64>
        %2 = stablehlo.multiply %cst_1, %arg0 : tensor<f64>
        %3 = stablehlo.multiply %2, %iterArg : tensor<f64>
        enzyme.yield %3 : tensor<f64>
      } attributes {activity = [#enzyme<activity enzyme_active>], ret_activity = [#enzyme<activity enzyme_activenoneed>]} : (tensor<f64>, tensor<f64>) -> tensor<f64>
      stablehlo.return %1 : tensor<f64>
    }
    return %0 : tensor<f64>
  }
}
```
Output
```
$ ./enzymexlamlir-opt test.mlir --outline-enzyme-regions --enzyme --canonicalize --remove-unnecessary-enzyme-ops --arith-raise
module @root {
  func.func @main() -> tensor<f64> {
    %cst = stablehlo.constant dense<1.000000e+00> : tensor<f64>
    %cst_0 = stablehlo.constant dense<2.000000e-01> : tensor<f64>
    %0 = stablehlo.while(%iterArg = %cst_0) : tensor<f64>
    cond {
      %1 = stablehlo.compare LT, %iterArg, %cst : (tensor<f64>, tensor<f64>) -> tensor<i1>
      stablehlo.return %1 : tensor<i1>
    } do {
      %1 = func.call @diffemain_to_diff0(%iterArg, %cst) : (tensor<f64>, tensor<f64>) -> tensor<f64>
      stablehlo.return %1 : tensor<f64>
    }
    return %0 : tensor<f64>
  }
  func.func @main_to_diff0(%arg0: tensor<f64>) -> tensor<f64> {
    %cst = stablehlo.constant dense<3.000000e+00> : tensor<f64>
    %0 = stablehlo.multiply %cst, %arg0 : tensor<f64>
    %1 = stablehlo.multiply %0, %arg0 : tensor<f64>
    return %1 : tensor<f64>
  }
  func.func private @diffemain_to_diff0(%arg0: tensor<f64>, %arg1: tensor<f64>) -> tensor<f64> {
    %cst = stablehlo.constant dense<3.000000e+00> : tensor<f64>
    %cst_0 = stablehlo.constant dense<0.000000e+00> : tensor<f64>
    %0 = stablehlo.multiply %cst, %arg0 : tensor<f64>
    %1 = stablehlo.add %arg1, %cst_0 : tensor<f64>
    %2 = stablehlo.multiply %1, %arg0 : tensor<f64>
    %3 = stablehlo.add %2, %cst_0 : tensor<f64>
    %4 = stablehlo.multiply %1, %0 : tensor<f64>
    %5 = stablehlo.add %4, %cst_0 : tensor<f64>
    %6 = stablehlo.multiply %3, %cst : tensor<f64>
    %7 = stablehlo.add %5, %6 : tensor<f64>
    return %7 : tensor<f64>
  }
}
```
We can compare this to
```
>>> jax.lax.while_loop(lambda x: x < 1.0, lambda x: jax.grad(lambda y: 3.0 * y * x)(x), 0.2)
Array(1.8000001, dtype=float32, weak_type=True)
```
which I take to mean `grad` treats `x` as constant wrt `y` such that the gradient is `3.0 * x` (using enzyme for this yields 1.2 i.e. `3.0 * 2.0 * 0.2`)

**Comment by Pangoraw** [comment]
this seems to be a bug in the region outlining:

```mlir
// $ cat while_error_grad.mlir
module @root {
  func.func @main(%arg1: tensor<f64>) -> tensor<f64> {
    %cst_0 = stablehlo.constant dense<1.000000e+00> : tensor<f64>
    %1 = enzyme.autodiff_region(%arg1, %cst_0) {
      ^bb0(%arg0: tensor<f64>):
        %cst_1 = stablehlo.constant dense<3.000000e+00> : tensor<f64>
         %2 = stablehlo.multiply %cst_1, %arg0 : tensor<f64>
         %3 = stablehlo.multiply %2, %arg1 : tensor<f64>
         enzyme.yield %3 : tensor<f64>
    } attributes {activity = [#enzyme<activity enzyme_active>], ret_activity = [#enzyme<activity enzyme_activenoneed>]} : (tensor<f64>, tensor<f64>) -> tensor<f64>
    return %1 : tensor<f64>
  }
}
// $ ./bazel-bin/enzymexlamlir-opt while_error_grad.mlir --outline-enzyme-regions
module @root {
  func.func @main(%arg0: tensor<f64>) -> tensor<f64> {
    %cst = stablehlo.constant dense<1.000000e+00> : tensor<f64>
    %0 = enzyme.autodiff @main_to_diff0(%arg0, %cst) {activity = [#enzyme<activity enzyme_active>], ret_activity = [#enzyme<activity enzyme_activenoneed>]} : (tensor<f64>, tensor<f64>) -> tensor<f64>
    return %0 : tensor<f64>
  }
  func.func private @main_to_diff0(%arg0: tensor<f64>) -> tensor<f64> {
    %cst = stablehlo.constant dense<3.000000e+00> : tensor<f64>
    %0 = stablehlo.multiply %cst, %arg0 : tensor<f64>
    %1 = stablehlo.multiply %0, %arg0 : tensor<f64>
    return %1 : tensor<f64>
  }
}
```

cc @vimarsh6739, i don't have full context on https://github.com/EnzymeAD/Enzyme/pull/2559

**Comment by vimarsh6739** [comment]
[That](https://github.com/EnzymeAD/Enzyme/pull/2559) basically ensures that we do not re-add the same primal argument as a const value when outlining to a standalone function. Let me take a look at this.
